mesh-design.md

```diff
@@ -0,0 +1,270 @@
+collider API proposal is designed primarily for use with `Actor.CreateEmpty`. See
+[Other Considerations](#removing-createprimitive-and-createfromgltf).
+
+Actor-creation-time collider initialization (1 and 2) is very convenient, because most of the time you don't
+need to change colliders at runtime. The challenge is that with mesh management, the mesh may not yet be loaded
+when the actor is created, or the mesh may be assigned well after actor creation. To match the convenience of the
+old mesh actor creation methods, I propose two changes.
+
+First, the creation of a new collider type: `auto`. When a mesh is loaded, we will register that mesh's preferred
+collider type based on the creation method, so primitives will get basic colliders (box, sphere, etc.) and glTF
+meshes will get either a bounding box or a mesh collider depending on `AssetContainer.loadGltf`'s `colliderType`
+argument (which matches the `CreateFromGltf` options). Then when a mesh is assigned to an actor with an `auto`
+collider type, a new collider is created on the actor based on the registered collider description for that mesh.
+
+Second, allow collider descriptions to be specified explicitly on initialization in the actor patch, instead of
+being a separate function argument like in `CreateFromGltf` or `CreatePrimitive`. This approach was rejected
+initially to prevent the user from specifying a mesh collider, but that could be done in other ways without
+complicating the API. See [Other Considerations](#consolidating-the-collider-apis-into-actor-patch).
+
+For example, if I wanted to create a box button, I might do something like this:
+
+```ts
+const context: MRE.Context;
+const assets: MRE.AssetContainer;
+
+// With a mesh asset and auto collider:
+const actor = MRE.Actor.CreateEmpty(context, {
+	actor: {
+		name: 'button',
+		appearance: {
+			meshId: assets.createBoxMesh('button', 0.25, 0.25, 0.1).id
+		},
+		// Note: This is proposed syntax. Today you'd have to call setCollider after creation.
+		// I.e. actor.setCollider('auto')
+		collider: { geometry: { shape: 'auto' } as MRE.AutoColliderGeometry }
+	}
+});
+
+// This would be equivalent to the following without `auto`:
+actor = MRE.Actor.CreateEmpty(context, {
+	actor: {
+		name: 'button',
+		appearance: {
+			meshId: assets.createBoxMesh('button', 0.25, 0.25, 0.1).id
+		}
+	}
+});
+actor.setCollider('box', false, { x: 0, y: 0, z: 0 }, { x: 0.25, y: 0.25, z: 0.1 });
+
+// Or this without mesh assets:
+actor = MRE.Actor.CreatePrimitive(context, {
+	definition: {
+		shape: 'box',
+		dimensions: { x: 0.25, y: 0.25, z: 0.1 }
+	},
+	addCollider: true,
+	actor: { name: 'button' }
+});
+```
+
+## Other Considerations
+
+### Consolidating the collider APIs into actor patch
+
+As previously noted, `CreatePrimitive` has one way to specify a collider, `CreateFromGltf` has another,
+and `setCollider` has yet another. This setup was done specifically to make it difficult to change actor
```

Comment:
I assume this is leaving in place the setCollider API calls on the actor for adding a collider to an actor after creation has happened?

Reply:
Correct, to match the other components that go through the patch: initialization through the Like interfaces, and modification through the dedicated API.

---
In reply to: [311278852](https://github.com/microsoft/mixed-reality-extension-sdk/pull/409#discussion_r311278852) [](ancestors = 311278852)